lsif/src/cache.test.ts

```diff
@@ -0,0 +1,145 @@
+import { GenericCache } from './cache'
+import * as sinon from 'sinon'
+
+describe('GenericCache', () => {
+    it('should evict items based by reverse recency', async () => {
+        const values = [
+            'foo', // foo*
+            'bar', // bar* foo
+            'baz', // baz* bar foo
+            'bonk', // bonk* baz bar foo
+            'quux', // quux* bonk baz bar foo
+            'bar', // bar quux bonk baz foo
+            'foo', // foo bar quux bonk baz
+            'honk', // honk* foo bar quux bonk
+            'foo', // foo honk bar quux bonk
+            'baz', // baz* foo honk bar quux
+        ]
+
+        // These are the cache values that need to be created, in-order
+        const expectedInstantiations = ['foo', 'bar', 'baz', 'bonk', 'quux', 'honk', 'baz']
+
+        let i = 0
+        const factory = sinon.stub()
```

Comment:
`stub()` cannot infer parameter and return types, so they default to `any[]`/`any`. You need to specify them manually to get type safety:
```suggestion
        const factory = sinon.stub<[string], Promise<string>>()
```

(note this doesn't apply to `spy()` because it can infer the types from the passed function, which is why it is often the better choice. But here `stub()` makes sense)

Reply:
What confidence does this give you in a test? I can build it with any generic types at all here and I can't get it to reject the program.

Do we really require type safety in our stubs/mocks in unit tests?

Reply:
Added type params to sinon.stubs.

Reply:
It means if you or someone else less familiar with the code refactors the code and changes the signature, TS will point out the tests that need to be adjusted. We were using `jest.fn()` a lot in the browser extension / extension API codebase, which doesn't have this type safety compare to sinon, and it made refactoring a lot more difficult.